HPCC-19393 Implement single-use security manager plugin

rpastrana opened the pull request:
- User/pass provided in config by the espcfg component
- User/pass must be provided to ESP.exe via command line params
- Defaults user to "admin" if omitted

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

rpastrana commented:
@wangkx please review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: cdf35cd0759f56ba0f811ff4de2efc83c19b38f0
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    5 sec (00:00:05) |  186 sec (00:03:06) |   55 sec (00:00:55) |    5 sec (00:00:05) |   22 sec (00:00:22) | 1112 sec (00:18:32) |   17 sec (00:00:17) | 1402 sec (00:23:22) |

wangkx commented on system/security/plugins/singleuserSecurity/singleUserSecurity.cpp:
```diff
@@ -0,0 +1,216 @@
+public:
+    CSingleUserSecurityManager(const char *serviceName, IPropertyTree *secMgrCfg, IPropertyTree *bindConfig) : CBaseSecurityManager(serviceName, (IPropertyTree *)NULL)
+    {
+        if (secMgrCfg)
```
Should you throw exception if secMgrCfg is nullptr? If secMgrCfg is nullptr, m_userPass will not be set and the authorize() will throw exception. So, why not throw exception here if secMgrCfg is nullptr?

rpastrana replied:
I agree, the problem is we tend not to throw in a constructor

wangkx commented on system/security/plugins/singleuserSecurity/singleUserSecurity.hpp:
```diff
@@ -0,0 +1,36 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2013 HPCC Systems®.
```
2018

wangkx commented on esp/platform/espcfg.cpp:
```diff
@@ -281,10 +282,41 @@ CEspConfig::CEspConfig(IProperties* inputs, IPropertyTree* envpt, IPropertyTree*
+            StringBuffer espuserpass;
+            m_inputs->getProp("SingleUsePass", espuserpass);
+            StringBuffer xpath;
+            xpath.setf("SecurityManagers/SecurityManager[@type=\"SingleUserSecurityManager\"]SingleUserSecurityManager/");
```
?

wangkx reviewed (commented):
@rpastrana most of my comments are minors.

rpastrana commented:
@wangkx please review latest commit.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: a0f52b8f270122e17d08cbf8457acdf8b74063bd
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   13 sec (00:00:13) |  181 sec (00:03:01) |   55 sec (00:00:55) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1092 sec (00:18:12) |   15 sec (00:00:15) | 1384 sec (00:23:04) |

wangkx commented on system/security/plugins/singleuserSecurity/CMakeLists.txt:
```diff
@@ -38,8 +38,8 @@ include_directories (
-         ${APR_INCLUDE_DIR}
-         ${APRUTIL_INCLUDE_DIR}
+         #${APR_INCLUDE_DIR}
+         #${APRUTIL_INCLUDE_DIR}
```
Why not remove those 2 lines?

wangkx commented on system/security/plugins/singleuserSecurity/CMakeLists.txt:
```diff
@@ -49,8 +49,8 @@ install ( TARGETS singleUserSecurity RUNTIME DESTINATION ${EXEC_DIR} LIBRARY DES
-         ${APR_LIBRARIES}
-         ${APRUTIL_LIBRARIES}
+         #${APR_LIBRARIES}
+         #${APRUTIL_LIBRARIES}
```
Why not remove those 2 lines?

wangkx commented on system/security/plugins/singleuserSecurity/singleUserSecurity.cpp:
```diff
@@ -0,0 +1,216 @@
+        CAuthMap* feature_authmap = new CAuthMap(this);
+
+        Owned<IPropertyTreeIterator> feature_iter;
+        feature_iter.setown(authconfig->getElements(".//Feature"));
```
Forgot this comment?

wangkx commented on system/security/plugins/singleuserSecurity/singleUserSecurity.cpp:
```diff
@@ -0,0 +1,216 @@
+        ForEach(*feature_iter)
+        {
+            IPropertyTree *feature = NULL;
+            feature = &feature_iter->query();
```
Forgot this one?

wangkx reviewed (commented):
@rpastrana back to you. For some reason, my 2 comments on system/security/plugins/singleuserSecurity/singleUserSecurity.cpp (Line 102 and Line 106) are not shown here. Please check.

rpastrana commented:
@wangkx one more review please

wangkx reviewed (approved):
@rpastrana looks good.

rpastrana commented:
@richardkchapman please merge

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: e6601d01aeecef938ce91c82cea4ecac3753ce46
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   13 sec (00:00:13) |  184 sec (00:03:04) |   56 sec (00:00:56) |    5 sec (00:00:05) |   21 sec (00:00:21) | 1109 sec (00:18:29) |   16 sec (00:00:16) | 1404 sec (00:23:24) |

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: e2dcc7723870a8027ae3c92392babf9a865b1cde
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   33 sec (00:00:33) |  182 sec (00:03:02) |   55 sec (00:00:55) |    7 sec (00:00:07) |   22 sec (00:00:22) | 1116 sec (00:18:36) |   16 sec (00:00:16) | 1431 sec (00:23:51) |

rpastrana committed:
HPCC-19393 Implement single-use security manager plugin

- User/pass must be provided to ESP.exe via command line params
- User/pass provided in configtree by the espcfg component
- ESPCfg encrypts password
- Defaults user to "admin" if omitted
- Fail to load if initial password is not provided

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 86e1f256b32113da507111f6f649b67a4b945024
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   17 sec (00:00:17) |  181 sec (00:03:01) |   55 sec (00:00:55) |    6 sec (00:00:06) |   22 sec (00:00:22) | 1108 sec (00:18:28) |   16 sec (00:00:16) | 1405 sec (00:23:25) |

rpastrana commented:
@wangkx please review. Added pass encryption. thanks.

wangkx reviewed (commented):
@rpastrana minor: 2 2016 s

rpastrana commented:
@wangkx last code review please

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: aea1c0bd50091c7213a46cafedc7beed506e0500
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   24 sec (00:00:24) |  183 sec (00:03:03) |   55 sec (00:00:55) |    5 sec (00:00:05) |   21 sec (00:00:21) | 1115 sec (00:18:35) |   16 sec (00:00:16) | 1419 sec (00:23:39) |

wangkx commented:
@rpastrana looks good.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: b1347f39450a3324c28237650d28d8b2b643cac1
Build: success
Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |83 |83 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |759 |759 |0 |
|test (thor) |686 |686 |0 |
|test (roxie) |799 |799 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   18 sec (00:00:18) |  181 sec (00:03:01) |   55 sec (00:00:55) |    6 sec (00:00:06) |   21 sec (00:00:21) | 1112 sec (00:18:32) |   16 sec (00:00:16) | 1409 sec (00:23:29) |
